File: algo/src/test/java/com/hartwig/actin/algo/evaluation/composite/NotTest.java
package com.hartwig.actin.algo.evaluation.composite;

import static com.hartwig.actin.algo.evaluation.EvaluationAssert.assertEvaluation;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;

import com.hartwig.actin.PatientRecord;
import com.hartwig.actin.TestDataFactory;
import com.hartwig.actin.algo.datamodel.Evaluation;
import com.hartwig.actin.algo.datamodel.EvaluationResult;
import com.hartwig.actin.algo.evaluation.TestEvaluationFunctionFactory;

import org.junit.Test;

public class NotTest {

    private static final PatientRecord TEST_PATIENT = TestDataFactory.createProperTestPatientRecord();

    @Test
    public void canNegateEvaluation() {
        assertEvaluation(EvaluationResult.FAIL, new Not(TestEvaluationFunctionFactory.pass()).evaluate(TEST_PATIENT));
        assertEvaluation(EvaluationResult.PASS, new Not(TestEvaluationFunctionFactory.fail()).evaluate(TEST_PATIENT));
        assertEvaluation(EvaluationResult.WARN, new Not(TestEvaluationFunctionFactory.warn()).evaluate(TEST_PATIENT));
        assertEvaluation(EvaluationResult.UNDETERMINED, new Not(TestEvaluationFunctionFactory.undetermined()).evaluate(TEST_PATIENT));
        assertEvaluation(EvaluationResult.NOT_IMPLEMENTED, new Not(TestEvaluationFunctionFactory.notImplemented()).evaluate(TEST_PATIENT));
        assertEvaluation(EvaluationResult.NOT_EVALUATED, new Not(TestEvaluationFunctionFactory.notEvaluated()).evaluate(TEST_PATIENT));
    }

    @Test
    public void canFlipMessagesAndMolecularEventsForPass() {
        Evaluation passed = CompositeTestFactory.create(EvaluationResult.PASS, true);

        Evaluation result = new Not(record -> passed).evaluate(TEST_PATIENT);

        assertEquals(result.recoverable(), passed.recoverable());

        assertFalse(passed.inclusionMolecularEvents().isEmpty());
        assertEquals(passed.inclusionMolecularEvents(), result.exclusionMolecularEvents());
        assertFalse(passed.exclusionMolecularEvents().isEmpty());
        assertEquals(passed.exclusionMolecularEvents(), result.inclusionMolecularEvents());

        assertEquals(passed.passSpecificMessages(), result.failSpecificMessages());
        assertEquals(passed.passGeneralMessages(), result.failGeneralMessages());
        assertEquals(passed.failSpecificMessages(), result.passSpecificMessages());
        assertEquals(passed.failGeneralMessages(), result.passGeneralMessages());

        assertEquals(passed.undeterminedSpecificMessages(), result.undeterminedSpecificMessages());
        assertEquals(passed.undeterminedGeneralMessages(), result.undeterminedGeneralMessages());
        assertEquals(passed.warnSpecificMessages(), result.warnSpecificMessages());
        assertEquals(passed.warnGeneralMessages(), result.warnGeneralMessages());
    }

    @Test
    public void canFlipMessagesAndMolecularEventsForFail() {
        Evaluation failed = CompositeTestFactory.create(EvaluationResult.FAIL, true);

        Evaluation result = new Not(record -> failed).evaluate(TEST_PATIENT);

        assertEquals(result.recoverable(), failed.recoverable());

        assertFalse(failed.inclusionMolecularEvents().isEmpty());
        assertEquals(failed.inclusionMolecularEvents(), result.exclusionMolecularEvents());
        assertFalse(failed.exclusionMolecularEvents().isEmpty());
        assertEquals(failed.exclusionMolecularEvents(), result.inclusionMolecularEvents());

        assertEquals(failed.passSpecificMessages(), result.failSpecificMessages());
        assertEquals(failed.passGeneralMessages(), result.failGeneralMessages());
        assertEquals(failed.failSpecificMessages(), result.passSpecificMessages());
        assertEquals(failed.failGeneralMessages(), result.passGeneralMessages());

        assertEquals(failed.undeterminedSpecificMessages(), result.undeterminedSpecificMessages());
        assertEquals(failed.undeterminedGeneralMessages(), result.undeterminedGeneralMessages());
        assertEquals(failed.warnSpecificMessages(), result.warnSpecificMessages());
        assertEquals(failed.warnGeneralMessages(), result.warnGeneralMessages());
    }

    @Test
    public void canRetainMessagesAndMolecularEventsForUndetermined() {
        Evaluation undetermined = CompositeTestFactory.create(EvaluationResult.UNDETERMINED, true);

        Evaluation result = new Not(record -> undetermined).evaluate(TEST_PATIENT);

        assertEquals(result.recoverable(), undetermined.recoverable());

        assertFalse(undetermined.inclusionMolecularEvents().isEmpty());
        assertEquals(undetermined.inclusionMolecularEvents(), result.inclusionMolecularEvents());
        assertFalse(undetermined.exclusionMolecularEvents().isEmpty());
        assertEquals(undetermined.exclusionMolecularEvents(), result.exclusionMolecularEvents());

        assertEquals(undetermined.passSpecificMessages(), result.passSpecificMessages());
        assertEquals(undetermined.passGeneralMessages(), result.passGeneralMessages());
        assertEquals(undetermined.failSpecificMessages(), result.failSpecificMessages());
        assertEquals(undetermined.failGeneralMessages(), result.failGeneralMessages());

        assertEquals(undetermined.undeterminedSpecificMessages(), result.undeterminedSpecificMessages());
        assertEquals(undetermined.undeterminedGeneralMessages(), result.undeterminedGeneralMessages());
        assertEquals(undetermined.warnSpecificMessages(), result.warnSpecificMessages());
        assertEquals(undetermined.warnGeneralMessages(), result.warnGeneralMessages());
    }
}

File: algo/src/test/kotlin/com/hartwig/actin/algo/evaluation/composite/NotTest.kt
package com.hartwig.actin.algo.evaluation.composite

import com.hartwig.actin.algo.evaluation.EvaluationAssert.assertEvaluation
import com.hartwig.actin.algo.evaluation.EvaluationFunction
import com.hartwig.actin.algo.evaluation.TestEvaluationFunctionFactory
import com.hartwig.actin.datamodel.PatientRecord
import com.hartwig.actin.datamodel.TestPatientFactory
import com.hartwig.actin.datamodel.algo.Evaluation
import com.hartwig.actin.datamodel.algo.EvaluationResult
import org.assertj.core.api.Assertions.assertThat
import org.junit.Test

class NotTest {

    @Test
    fun `Should correctly negate evaluation result`() {
        assertEvaluation(EvaluationResult.FAIL, Not(TestEvaluationFunctionFactory.pass()).evaluate(TEST_PATIENT))
        assertEvaluation(EvaluationResult.PASS, Not(TestEvaluationFunctionFactory.fail()).evaluate(TEST_PATIENT))
        assertEvaluation(EvaluationResult.WARN, Not(TestEvaluationFunctionFactory.warn()).evaluate(TEST_PATIENT))
        assertEvaluation(EvaluationResult.UNDETERMINED, Not(TestEvaluationFunctionFactory.undetermined()).evaluate(TEST_PATIENT))
        assertEvaluation(EvaluationResult.NOT_EVALUATED, Not(TestEvaluationFunctionFactory.notEvaluated()).evaluate(TEST_PATIENT))
    }

    @Test
    fun `Should flip messages and molecular events for pass evaluation`() {
        val passFunction = CompositeTestFactory.create(EvaluationResult.PASS, includeMolecular = true)
        val passed = passFunction.evaluate(TEST_PATIENT)
        val result: Evaluation = Not(passFunction).evaluate(TEST_PATIENT)
        assertMessagesAreFlipped(passed, result)
        assertEventsAreFlipped(passed, result)
    }

    @Test
    fun `Should flip messages and molecular events for fail evaluation`() {
        val failFunction = CompositeTestFactory.create(EvaluationResult.FAIL, includeMolecular = true)
        val failed = failFunction.evaluate(TEST_PATIENT)
        val result: Evaluation = Not(failFunction).evaluate(TEST_PATIENT)
        assertMessagesAreFlipped(failed, result)
        assertEventsAreFlipped(failed, result)
    }

    @Test
    fun `Should retain messages and flip molecular events for undetermined evaluation`() {
        val undeterminedFunction = CompositeTestFactory.create(EvaluationResult.UNDETERMINED, includeMolecular = true)
        val undetermined = undeterminedFunction.evaluate(TEST_PATIENT)
        val result: Evaluation = Not(undeterminedFunction).evaluate(TEST_PATIENT)
        assertMessagesAreRetained(undetermined, result)
        assertEventsAreFlipped(undetermined, result)
    }

    @Test
    fun `Should retain messages and flip molecular events for warn evaluation`() {
        val warnFunction = CompositeTestFactory.create(EvaluationResult.WARN, includeMolecular = true)
        val warn = warnFunction.evaluate(TEST_PATIENT)
        val result: Evaluation = Not(warnFunction).evaluate(TEST_PATIENT)
        assertMessagesAreRetained(warn, result)
        assertEventsAreFlipped(warn, result)
    }

    @Test
    fun `Should flip messages and molecular events for not evaluated evaluation`() {
        val notEvaluatedEvaluation = CompositeTestFactory.create(EvaluationResult.NOT_EVALUATED, includeMolecular = true)
        val notEvaluated = notEvaluatedEvaluation.evaluate(TEST_PATIENT)
        val result: Evaluation = Not(notEvaluatedEvaluation).evaluate(TEST_PATIENT)
        assertMessagesAreFlipped(notEvaluated, result)
        assertEventsAreFlipped(notEvaluated, result)
    }

    @Test
    fun `Should retain isMissingGenesForSufficientEvaluation value`() {
        val function: EvaluationFunction = CompositeTestFactory.create(isMissingGenes = true)
        val result: Evaluation = Not(function).evaluate(TEST_PATIENT)
        assertThat(result.isMissingGenesForSufficientEvaluation).isTrue()
    }

    private fun assertMessagesAreFlipped(evaluation: Evaluation, negatedEvaluation: Evaluation) {
        assertThat(evaluation.recoverable).isEqualTo(negatedEvaluation.recoverable)
        assertThat(negatedEvaluation.passMessages).isEqualTo(evaluation.failMessages)
        assertThat(negatedEvaluation.failMessages).isEqualTo(evaluation.passMessages)
        assertThat(negatedEvaluation.undeterminedMessages).isEqualTo(evaluation.undeterminedMessages)
        assertThat(negatedEvaluation.warnMessages).isEqualTo(evaluation.warnMessages)
    }

    private fun assertMessagesAreRetained(evaluation: Evaluation, negatedEvaluation: Evaluation) {
        assertThat(evaluation.recoverable).isEqualTo(negatedEvaluation.recoverable)
        assertThat(negatedEvaluation.passMessages).isEqualTo(evaluation.passMessages)
        assertThat(negatedEvaluation.failMessages).isEqualTo(evaluation.failMessages)
        assertThat(negatedEvaluation.undeterminedMessages).isEqualTo(evaluation.undeterminedMessages)
        assertThat(negatedEvaluation.warnMessages).isEqualTo(evaluation.warnMessages)
    }

    private fun assertEventsAreFlipped(evaluation: Evaluation, negatedEvaluation: Evaluation) {
        assertThat(evaluation.inclusionMolecularEvents).isNotEmpty()
        assertThat(negatedEvaluation.inclusionMolecularEvents).isEqualTo(evaluation.exclusionMolecularEvents)
        assertThat(evaluation.exclusionMolecularEvents).isNotEmpty()
        assertThat(negatedEvaluation.exclusionMolecularEvents).isEqualTo(evaluation.inclusionMolecularEvents)
    }

    companion object {
        private val TEST_PATIENT: PatientRecord = TestPatientFactory.createProperTestPatientRecord()
    }
}